Pulp interface xpress py

djunglas opened the pull request:
This pull request adds a new solver `XPRESS_PY`. Unlike the existing `XPRESS` solver, the new solver does not invoke an external process (the Xpress command line tool) but instead uses the Xpress Python API directly.

This is more flexible and gives access to more features (in fact all) of the Xpress optimizer.

We also introduce an alias `XPRESS_CMD` for the existing `XPRESS` solver to meet the naming conventions that were established for other solvers (`CMD` goes through an external command, `PY` uses the solver's Python API).

pchtsp commented on pulp/apis/xpress_api.py:
```diff
@@ -318,3 +325,424 @@ def quote_path(path):
+        except (xpress.ModelError, xpress.InterfaceError, xpress.SolverError) as err:
+            raise PulpSolverError(str(err))
+
+    def actualSolve(self, lp, prepare=None):
+        """Solve a well formulated lp problem"""
+        if not self.available():
+            # Import again to get a more verbose error message
+            message = "XPRESS Python API not available"
+            try:
+                import xpress
+            except ImportError as err:
+                message = str(err)
+            raise PulpSolverError(message)
+
+        self._extract(lp)
+        try:
+            model = lp._xprs
+            # Apply controls that were passed to the constructor
```
it would be nice to keep the `GUROBI`, `CPLEX_PY` convention of using `self.solverModel` to store the solver object. Also, I'd prefer to have the same convention of having the `actualSolve` that does a "simple solve" and then having the possibility (as documented [here](https://coin-or.github.io/pulp/guides/how_to_configure_solvers.html#using-solver-specific-functionality)) to split the solving into steps in order to tweak the solver object. The steps should be: (1) build the solver model, (2) tweak it, (3) call the solver.

djunglas replied:
I changed the code as requested.
- Renamed `lp._xprs` to `lp.solverModel`
- Split the solution process into `buildSolverModel`, `callSolver`, `findSolutionValues`

pchtsp reviewed (commented):
Thanks for the PR!
I added some comments trying to standardize the interface so it looks like other existing python-interfaces for solvers.

Lastly, I think that having this solver would make it possible to add tests for this API in github actions? Check [here](https://github.com/fico-xpress/pulp/blob/master/.github/workflows/pythonpackage.yml) to see if we could add a `pip install xpress` or something in the tests flow as long as the python package would let us use it for small problems without a license.

pchtsp commented:
Hello again! do you think you could do the changes in the near future? I was thinking on releasing a new version of pulp soon and it would be nice to include this new solver. Thanks!

djunglas commented:
> Hello again! do you think you could do the changes in the near future? I was thinking on releasing a new version of pulp soon and it would be nice to include this new solver. Thanks!

I'll try to make the requested changes today.

djunglas commented on pulp/apis/xpress_api.py:
```diff
@@ -318,3 +325,424 @@ def quote_path(path):
+    name = "XPRESS_CMD"
+
+    def __init__(self, *args, **kwargs):
+        XPRESS.__init__(self, *args, **kwargs)
```
I changed to `XPRESS_CMD = XPRESS`.

djunglas commented on pulp/apis/xpress_api.py:
```diff
@@ -108,7 +117,7 @@ def available(self):
         return self.executable(self.path)
 
-    def actualSolve(self, lp):
+    def actualSolve(self, lp, prepare=None):
```
Removed.

djunglas commented:
> Thanks for the PR! I added some comments trying to standardize the interface so it looks like other existing python-interfaces for solvers.
> 
> Lastly, I think that having this solver would make it possible to add tests for this API in github actions? Check [here](https://github.com/fico-xpress/pulp/blob/master/.github/workflows/pythonpackage.yml) to see if we could add a `pip install xpress` or something in the tests flow as long as the python package would let us use it for small problems without a license.

@pchtsp, I think I addressed all your comments. About adding something to github actions: I think that should be possible and `pip install xpress` should be sufficient. The default license that comes with the Xpress Python package should allow all the small models to be solved. However, I am not familiar with github actions and would rather leave that change (and testing it) to someone else (you?).

pchtsp commented:
thank you so much for all this work!
